cpp/src/arrow/compute/kernels/codegen_internal.h

```diff
@@ -349,6 +356,31 @@ struct ArrayIterator<Type, enable_if_base_binary<Type>> {
   }
 };
 
+template <typename Type>
+struct ArrayIterator<Type, enable_if_list_type<Type>> {
+  using offset_type = typename Type::offset_type;
+
+  const ArraySpan& arr;
+  const offset_type* offsets;
+  offset_type cur_offset;
+  int64_t position;
+
+  explicit ArrayIterator(const ArraySpan& arr)
+      : arr(arr),
+        offsets(reinterpret_cast<const offset_type*>(arr.buffers[1].data) + arr.offset),
+        cur_offset(offsets[0]),
+        position(0) {}
+
+  ArraySpan operator()() {
+    offset_type next_offset = offsets[++position];
+    const offset_type length = next_offset - cur_offset;
+    ArraySpan new_span{arr};
+    new_span.SetSlice(cur_offset, length);
```

Comment:
I think this is a bit confused. This is creating a list ArraySpan with `length` list elements in it.

If you want to use `ArrayIterator`, I think it should create an ArraySpan of the child array.